middle_end/flambda2/from_lambda/closure_conversion.ml

```diff
@@ -1578,6 +1482,244 @@ let close_let_rec acc env ~function_declarations
+      ~handler:perform_over_application ~body ~is_exn_handler:false
+  in
+  match needs_region with
+  | None -> acc, both_applications
+  | Some (region, _) ->
+    Let_with_acc.create acc
+      (Bound_pattern.singleton (Bound_var.create region Name_mode.normal))
+      (Named.create_prim (Nullary Begin_region) apply_dbg)
+      ~body:both_applications
+    |> Expr_with_acc.create_let
+
+let close_apply acc env (apply : IR.apply) : Acc.t * Expr_with_acc.t =
+  let callee = find_simple_from_id env apply.func in
+  let approx = Env.find_value_approximation env callee in
+  let code_info =
+    match approx with
+    | Closure_approximation (_, _, Code_present code) ->
+      Some
+        ( Code.params_arity code,
+          Code.is_tupled code,
+          Code.num_trailing_local_params code,
+          Code.contains_no_escaping_local_allocs code,
+          Code.result_arity code )
+    | Closure_approximation (_, _, Metadata_only metadata) ->
+      Some
+        ( Code_metadata.params_arity metadata,
+          Code_metadata.is_tupled metadata,
+          Code_metadata.num_trailing_local_params metadata,
+          Code_metadata.contains_no_escaping_local_allocs metadata,
+          Code_metadata.result_arity metadata )
+    | Value_unknown -> None
+    | _ ->
+      if Flambda_features.check_invariants ()
+      then
+        Misc.fatal_errorf
+          "Unexpected approximation for callee %a in [Closure_conversion], \
+           expected a closure approximation."
+          Simple.print callee
+      else None
+  in
+  match code_info with
+  | None -> close_exact_or_unknown_apply acc env apply None
+  | Some
+      ( arity,
+        is_tupled,
+        num_trailing_local_params,
+        contains_no_escaping_local_allocs,
+        result_arity ) -> (
+    let args, missing_args, remaining_args =
+      let rec split args arity =
+        match args, arity with
+        | _, [] -> [], [], args
+        | [], _ -> [], arity, []
+        | arg :: args, _ :: arity ->
+          let args, missing, remains = split args arity in
+          arg :: args, missing, remains
+      in
+      let arity =
+        if is_tupled
+        then
+          Flambda_arity.With_subkinds.create
+            [Flambda_kind.With_subkind.block Tag.zero arity]
+        else arity
+      in
+      split apply.args arity
+    in
+    let exact_call apply_continuation acc =
+      match missing_args with
+      | [] ->
+        close_exact_or_unknown_apply acc env
+          { apply with args; continuation = apply_continuation }
+          (Some approx)
+      | _ :: _ ->
+        wrap_partial_application acc env apply_continuation apply approx args
+          missing_args ~arity ~num_trailing_local_params
+          ~contains_no_escaping_local_allocs
+    in
+    match remaining_args with
+    | [] -> exact_call apply.continuation acc
+    | args ->
+      wrap_over_application acc env exact_call apply args
+        ~contains_no_escaping_local_allocs ~result_arity)
```

Comment:
I think this could be refactored to a single match on both `remaining_args` and `missing_args`. It's a bit confusing to have an `exact_call` function that may not perform an exact call.

Reply:
Yes, but the full call shouldn't be dispatching on `missing_args`. In a way, `split` shoud return `Exact | Partial_app of missing_args | Over_app of remaining_args` in addition to the actual arguments. I don't think we can get in a case where both remaining args and missing args occur, and the code should reflect this better.